oak/server/rust/oak_runtime/src/channel.rs

```diff
@@ -0,0 +1,255 @@
+    fn is_empty_orphan(&self) -> bool {
+        self.0.is_orphan() && self.read(|messages| messages.len()) == 0
+    }
+}
+
+impl ChannelWriter {
+    /// Write a message to a channel. Fails with `OakStatus::ERR_CHANNEL_CLOSED` if the underlying
+    /// channel has been orphaned.
+    pub fn write(&self, msg: Message) -> Result<(), OakStatus> {
+        self.0.try_write(|messages| messages.push_back(msg))
+    }
+}
+
+impl Drop for ChannelWriter {
+    fn drop(&mut self) {
+        self.0.remove_writer();
+    }
+}
+
+impl Clone for ChannelWriter {
+    fn clone(&self) -> Self {
+        self.0.add_writer();
+        ChannelWriter(self.0.clone())
+    }
+}
+
+/// A helper type to determine if `try_read_message` was called with not enough `bytes_capacity`
+/// and/or `handles_capacity`.
+pub enum TakeMessage {
+    Success(Message),
+    NeedsCapacity(usize, usize),
+}
+
+impl ChannelReader {
+    /// Thread safe. Read a message from a channel. Fails with `OakStatus::ERR_CHANNEL_CLOSED` if
+    /// the underlying channel has been orphaned _and_ is empty.
+    pub fn read(&self) -> Result<Option<Message>, OakStatus> {
+        if self.0.is_empty_orphan() {
+            return Err(OakStatus::ERR_CHANNEL_CLOSED);
+        }
+        Ok(self.0.write(|messages| messages.pop_front()))
+    }
+
+    /// Thread safe. Returns `Ok(true)` if there is at least one message in the channel. Fails with
+    /// `OakStatus::ERR_CHANNEL_CLOSED` if the underlying channel has been orphaned _and_ is empty.
+    pub fn has_message(&self) -> Result<bool, OakStatus> {
+        if self.0.is_empty_orphan() {
+            return Err(OakStatus::ERR_BAD_HANDLE);
+        }
+        Ok(self.0.read(|messages| !messages.is_empty()))
+    }
+
+    /// Thread safe. Reads a message from the channel if `bytes_capacity` and `handles_capacity` are
+    /// large enough to accept the message. Fails with `OakStatus::ERR_CHANNEL_CLOSED` if the
+    /// underlying channel has been orphaned _and_ is empty. If there was not enough
+    /// `bytes_capacity` or `handles_capacity`, `try_read_message` will return
+    /// `Some(TakeMessage::NeedsCapacity(needed_bytes_capacity,needed_handles_capacity))`. Does not
+    /// guarantee that the next call will succeed after capacity adjustments as another thread may
+    /// have read the original message.
+    pub fn try_read_message(
+        &self,
+        bytes_capacity: usize,
+        handles_capacity: usize,
+    ) -> Result<Option<TakeMessage>, OakStatus> {
+        if self.0.is_empty_orphan() {
+            return Err(OakStatus::ERR_CHANNEL_CLOSED);
+        }
+        Ok(self.0.write(|messages| {
+            let front = messages.front()?;
+
+            let req_bytes_capacity = front.data.len();
+            let req_handles_capacity = front.channels.len();
+
+            if req_bytes_capacity > bytes_capacity || req_handles_capacity > handles_capacity {
+                return Some(TakeMessage::NeedsCapacity(
```

Comment:
Could inline the `else` branch of the if and remove the explicit `return` here.

Reply:
Done